### PR TITLE
netsync: Split header sync from block sync.

### PR DESCRIPTION
**Pull request by davecgh**
This refactors the code that deals with starting synchronization with the sync peer to split out the pieces specific to the initial header sync from the chain data sync.  Along the way it removes some duplicate checks that are no longer necessary.

The primary motivation is to help pave the way for eventually performing parallel block downloads, but it also has the added benefit of making the logic a bit easier to follow, so it is useful on its own.

### CODE REVIEW

File: internal/netsync/manager.go [hunk 1]
```diff
@@ -474,98 +474,96 @@ func (m *SyncManager) fetchNextHeaders(peer *Peer) {
...
 
-	// Start the header sync progress stall timeout when the initial headers
-	// sync is not already done.
+	// Start the header sync progress stall timeout.
+	m.hdrSyncState.resetStallTimeout()
+}
+
+// startSync will choose the best peer among the available candidate peers to
+// download/sync the blockchain from.  It also examines the candidates for any
+// which are no longer candidates and removes them as needed.
```
Comment (matthawkins90) [inline internal/netsync/manager.go:542]: Same comment here as above. Peers are not being disconnected, so the "removed" wording is ambiguous/confusing.

Suggestion:

```
// startSync selects the peer with the most blocks and begins to download the
// blockchain headers from them. It also examines the sync candidates and marks
// those whose chain heights have been surpassed as ineligible to sync from.
```

**Review by matthawkins90 — COMMENTED** [review]
In addition to my comments about wording, I had one other thought. You said several times in the comments that the goal of these functions was to choose the "best" sync candidate, when really they're just choosing the candidate with the most blocks.

Let's say multiple peers have the blocks we need.
How difficult would it be to define "best" as:
* They have the blocks we need, and
* Peer A is faster than all the other peers that also have the blocks we need.

So even if Peers B and C may ultimately have "more" blocks, we can start with Peer A because it's fast, and then move to Peers B and/or C once Peer A has exhausted their usefulness.

**Comment by davecgh** [comment]
> Let's say multiple peers have the blocks we need. How difficult would it be to define "best" as:
> 
>     * They have the blocks we need, and
>     * Peer A is faster than all the other peers that also have the blocks we need.
> 
> So even if Peers B and C may ultimately have more blocks, we can start with Peer A because it's fast, and then move to Peers B and/or C once Peer A has exhausted their usefulness.

Indeed.  Although it isn't very obvious given the current structure, there are two different cases at play:

1) The initial header sync which doesn't take long and can't really be parallelized (at least not without significant protocol changes, that is) because all headers have to be proven to connect and every header requires the previous one it links to in order to do that.
2) The block sync based based on the headers.

So, the ultimate goal is to make it so that the sync peer is only used for the the first case, and get rid of the entire notion of it for the second case and instead download them all in parallel from whatever peers have what we need.

That is what I was referring to by "parallel block download" in the PR description.

File: internal/netsync/manager.go [hunk 2]
```diff
@@ -474,98 +474,96 @@ func (m *SyncManager) fetchNextHeaders(peer *Peer) {
...
-func (m *SyncManager) startSync() {
-	// Nothing more to do when already syncing.
-	if m.syncPeer != nil {
-		return
-	}
-
-	chain := m.cfg.Chain
-	best := chain.BestSnapshot()
+// updateSyncPeerState updates the sync peer to the best candidate and removes
+// any sync candidates that are no longer candidates.
```
[the remark this answers is not in the document]
Comment (davecgh) [reply internal/netsync/manager.go:478]: Thanks for the suggestion.  Good point about it being fairly easy to misinterpret remove in this context.  I'll use your wording as is.  It's nice and clear!

I was intentionally being a little bit vague with the "best" wording, because ultimately I'd like to have a fitness function that is used to determine the best candidate and "best" covers that.  However, the comments can always be updated as needed, so I agree it's better to be a little more explicit here.

File: internal/netsync/manager.go (hunk 1 above)
Comment (davecgh) [reply internal/netsync/manager.go:542]: Hmm.  I agree with the potential ambiguity around "removed", but the rest of the comment is questionable since it really does more than just the blockchain headers depending on the state of things.

Comment (davecgh) [reply internal/netsync/manager.go]: I updated it clarify a bit.

Comment (davecgh) [reply internal/netsync/manager.go]: I updated it clarify a bit.  Does the new text make things more clear?

Comment (matthawkins90) [reply internal/netsync/manager.go]: Yep, thanks for clarifying!

**Review by matthawkins90 — APPROVED** [review]

**Comment by matthawkins90** [comment]
Whoops, I see now that I was using the wrong terminology when I asked about switching to faster peers during the headers sync. Glad to see that there's still a possibility for a fitness function when choosing a peer to sync headers from. I understand that the headers download is typically ultrafast and needs to happen consecutively. I just did a test IBD on mainnet and it took me 47 seconds to download all 873,000 headers, so it's not like that part is the bottleneck. I guess I was just thinking about protecting for the edge-case where the peer with the most headers has a latency of 300ms but the peer with just a few fewer headers has a latency of 15ms. But yeah, probably not a bottleneck that's worth optimizing.

**Comment by davecgh** [comment]
>... I understand that the headers download is typically ultrafast and needs to happen consecutively. I just did a test IBD on mainnet and it took me 47 seconds to download all 873,000 headers, so it's not like that part is the bottleneck. I guess I was just thinking about protecting for the edge-case where the peer with the most headers has a latency of 300ms but the peer with just a few fewer headers has a latency of 15ms. But yeah, probably not a bottleneck that's worth optimizing.

It's perhaps also worth noting that there is already a failsafe against that type of behavior.  Namely, there is a header sync stall timeout that will cause it to choose a different peer to sync the headers from if the process stalls out too long.

In practice though, as you mentioned, it's not really a bottleneck.  Also, since it only happens once at initial startup, by the time you figure out which peer might be the fastest, it's probably already done.

**Review by jholdstock — APPROVED** [review]

**Review by dajohi — APPROVED** [review]
